### DISCUSSION TITLE
Minkowski algebra

### DISCUSSION THREAD

**Discussion by CaseySanchez** [laffernandes/gatl#32] [Q&A]
Hello Dr. Fernandes,

Do you have any examples using the Minkowski algebra?

Cheers,
Casey

**Comment by laffernandes** [comment]
No, I don't have. But I can help you tô prepare some.

**Reply by CaseySanchez** [reply]
I suppose that a helpful & simple example may be cloned from Enki's Ganja.js Coffeeshop (uses a `P=1, Q=3` signature instead of `P=3, Q=1` signature):

https://enkimute.github.io/ganja.js/examples/coffeeshop.html#timespace_lorentz

In your implementation, am I correct in presuming that `em` corresponds to the basis element of time? I am also interested in adding an additional `R=4` dimensions for the purposes of computing the spacetime gradient with auto-differentiation.

**Reply by CaseySanchez** [reply]
This appears to work:

```
#include <iostream>
#include <gatl/ga3m.hpp>

int main(int argc, char **argv)
{
    using namespace ga3m;
    
    auto micros = [](double t) {
        return t * 1e-6 * em;
    };

    auto km = [](double x, double y, double z) {
        return (x * e1 + y * e2 + z * e3) / 299792.458;
    };
    
    auto strike_tree = micros(10.0) + km(0.0, 0.0, 0.0);
    auto strike_pole = micros(10.0) + km(20.0, 0.0, 0.0);
    
    auto rocket_speed = 0.5;
    auto rocket_frame = exp(std::atanh(rocket_speed) * 0.5 * (e1 ^ em));

    auto strike_tree_r = apply_even_versor(rocket_frame, strike_tree);
    auto strike_pole_r = apply_even_versor(rocket_frame, strike_pole);
    
    std::cout << "rocket sees lightning strike the tree at time = " << (*std::next(std::cbegin(strike_tree_r.values()), 3) * 1e+6) << "ms" << std::endl;
    std::cout << "rocket sees lightning strike the pole at time = " << (*std::next(std::cbegin(strike_pole_r.values()), 3) * 1e+6) << "ms" << std::endl;

    return 0;
}
```

Is there a better way to access the individual elements than using `values()`?
